### DISCUSSION TITLE
Tile serving example with FastAPI

### DISCUSSION THREAD

**Discussion by banesullivan** [banesullivan/server-thread#3] [Show and tell]
This is a simple example to show how easy dynamic tile serving can be with [`large-image`](https://github.com/girder/large_image) - this is effectively all that [`localtileserver`](https://github.com/banesullivan/localtileserver) does

```py
import io

from fastapi import FastAPI
from ipyleaflet import Map, TileLayer
import large_image
from starlette.responses import StreamingResponse

from server_thread import ServerThread

src = large_image.open(
    "/Users/bane/Desktop/rasters/TC_NG_SFBay_US_Geo.tif", projection="EPSG:3857", encoding="PNG"
)

app = FastAPI()


@app.get("/metadata")
async def get_metadata():
    return src.getMetadata()


@app.get("/tile/{z}/{x}/{y}.png")
async def get_tile(z: int, x: int, y: int):
    tile_binary = src.getTile(x, y, z)
    return StreamingResponse(io.BytesIO(tile_binary), media_type="image/png")


# Launch app in a background thread
server = ServerThread(app)

t = TileLayer(
    url=f"http://{server.host}:{server.port}/tile/{{z}}/{{x}}/{{y}}.png",
    attribution="booya!",
)

m = Map(zoom=8, center=[37.8751050766018, -122.19290725886823])
m.add_layer(t)
m
```
<img width="1068" alt="Screen Shot 2022-07-27 at 5 22 27 PM" src="https://user-images.githubusercontent.com/22067021/181389069-ac79f8c5-f6d8-46ab-a9ab-c25e99cb9650.png">

**Comment by vincentsarago** [comment]
that's really cool 🙏

**Reply by banesullivan** [reply]
Been playing with `rio-tiler` lately and put this little example together:

```py
import io
from fastapi import FastAPI
from rio_tiler.io import Reader
import numpy as np
from ipyleaflet import Map, TileLayer
from starlette.responses import StreamingResponse
from server_thread import ServerThread

path = 'image.tif'

src = Reader(path)


def tile(src, z, x, y):
    # return np.moveaxis(src.tile(x, y, z, ).data, 0, -1)
    class PNGBytes(bytes):
        def _repr_png_(self):
            return self
    return PNGBytes(src.tile(x, y, z, ).render(img_format="png"))


app = FastAPI()
@app.get("/tile/{z}/{x}/{y}.png")
async def get_tile(z: int, x: int, y: int):
    tile_binary = tile(src, z, x, y)
    return StreamingResponse(io.BytesIO(tile_binary), media_type="image/png")
# Launch app in a background thread
server = ServerThread(app)


m = Map(zoom=8, center=[37.8751050766018, -122.19290725886823])
t = TileLayer(
    url=f"http://{server.host}:{server.port}/tile/{{z}}/{{x}}/{{y}}.png",
    attribution="booya!",
)
m.add(t)
m
```
<img width="1107" alt="Screen Shot 2023-05-04 at 2 11 11 PM" src="https://user-images.githubusercontent.com/22067021/236318395-fc506faf-b254-472d-b0dc-b6b96bdb186e.png">

**Reply by banesullivan** [reply]
FYI, I've been playing around with Tornado in large-image, and with JupyterLab, we can avoid server-thread: https://github.com/girder/large_image/pull/1065

**Reply by vincentsarago** [reply]
nice, I'm using server-thread in rio-viz https://github.com/developmentseed/rio-viz/blob/main/rio_viz/app.py#L821-L840 

https://github.com/developmentseed/rio-viz/tree/main#in-notebook-environment

**Reply by banesullivan** [reply]
I really wish `ipyleaflet` could just accept a Python callable (e.g. the `tile()` method) and send the image binary contents via its comms

**Reply by banesullivan** [reply]
Oh that's awesome!! I hadn't seen rio-viz, will definitely try it out

**Reply by vincentsarago** [reply]
well we could also use geotiff js instead of python... but where is the fun of that !
